File: 81.search-in-rotated-sorted-array-ii.java
/*
 * @lc app=leetcode id=81 lang=java
 *
 * [81] Search in Rotated Sorted Array II
 */

// @lc code=start

class Solution {
    public boolean search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left < right) {
            int mid = (left + right) / 2;

            if (nums[mid] == target) {
                return true;
            } else if (nums[mid] > nums[right]) {
                // if left part is sorted
                if (nums[left] <= target && target < nums[mid]) {
                    // target cannot be equal to nums[mid]
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else if (nums[mid] < nums[right]) {
                // if right part is sorted
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            } else {
                /**
                 * duplicates, we know nums[pivot]=nums[left]=nums[right], but != target based
                 * on current information, we can only move left pointer to skip one cell thus
                 * in the worest case, we would have target: 2, and array like 11111111, then
                 * the running time would be O(n)
                 */
                right--;
            }
        }
        if (left == right && nums[left] == target)
            return true;
        else
            return false;
    }
}

// @lc code=end

File: 81.search-in-rotated-sorted-array-ii.py
#
# @lc app=leetcode id=81 lang=python3
#
# [81] Search in Rotated Sorted Array II
#

# @lc code=start

class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums)-1
        while l < r:
            mid = (l+r)//2

            if nums[mid] == target:
                return True
            elif nums[mid] > nums[r]:
                if nums[l] <= target and target < nums[mid]:
                    r = mid-1
                else:
                    l = mid+1
            elif nums[mid] < nums[r]:
                if nums[mid] < target and target <= nums[r]:
                    l = mid+1
                else:
                    r = mid-1
            else:
                r -= 1

        return l == r and nums[l] == target
# @lc code=end
